MoltenVK/MoltenVK/Utility/MVKVectorAllocator.h

```diff
@@ -0,0 +1,527 @@
+  {
+    a.ptr = nullptr;
+    a.num_elements_used = 0;
+    a.num_elements_reserved = 0;
+  }
+
+  ~mvk_vector_allocator_default()
+  {
+    deallocate();
+  }
+
+  size_t get_capacity() const
+  {
+    return num_elements_reserved;
+  }
+
+  void swap( mvk_vector_allocator_default &a )
+  {
+    const auto copy_ptr                   = a.ptr;
+    const auto copy_num_elements_used     = a.num_elements_used;
+    const auto copy_num_elements_reserved = a.num_elements_reserved;
+
+    a.ptr                   = ptr;
+    a.num_elements_used     = num_elements_used;
+    a.num_elements_reserved = num_elements_reserved;
+
+    ptr                   = copy_ptr;
+    num_elements_used     = copy_num_elements_used;
+    num_elements_reserved = copy_num_elements_reserved;
+  }
+
+  void allocate( const size_t num_elements_to_reserve )
+  {
+    deallocate();
+
+    ptr                   = reinterpret_cast< T* >( mvk_memory_allocator::alloc( num_elements_to_reserve * sizeof( T ) ) );
+    num_elements_used     = 0;
+    num_elements_reserved = num_elements_to_reserve;
+  }
+
+  void re_allocate( const size_t num_elements_to_reserve )
+  {
+    //if constexpr( std::is_trivially_copyable<T>::value )
+    //{
+    //  ptr = reinterpret_cast< T* >( mvk_memory_allocator::tm_memrealloc( ptr, num_elements_to_reserve * sizeof( T ) );
+    //}
+    //else
+    {
+      auto *new_ptr = reinterpret_cast< T* >( mvk_memory_allocator::alloc( num_elements_to_reserve * sizeof( T ) ) );
+
+      for( size_t i = 0; i < num_elements_used; ++i )
+      {
+        construct( &new_ptr[i], std::move( ptr[i] ) );
+        destruct( &ptr[i] );
+      }
+
+      //if ( ptr != nullptr )
+      {
+        mvk_memory_allocator::free( ptr );
+      }
+
+      ptr = new_ptr;
+    }
+
+    num_elements_reserved = num_elements_to_reserve;
+  }
+
+  void shrink_to_fit()
+  {
+    if( num_elements_used == 0 )
+    {
+      deallocate();
+    }
+    else
+    {
+      auto *new_ptr = reinterpret_cast< T* >( mvk_memory_allocator::alloc( num_elements_used * sizeof( T ) ) );
+
+      for( size_t i = 0; i < num_elements_used; ++i )
+      {
+        construct( &new_ptr[i], std::move( ptr[i] ) );
+        destruct( &ptr[i] );
+      }
+
+      mvk_memory_allocator::free( ptr );
+
+      ptr = new_ptr;
+      num_elements_reserved = num_elements_used;
+    }
+  }
+
+  void deallocate()
+  {
+    destruct_all<T>();
+
+    mvk_memory_allocator::free( ptr );
+
+    ptr = nullptr;
+    num_elements_reserved = 0;
+  }
+};
+
+
+//////////////////////////////////////////////////////////////////////////////////////////
+//
+// mvk_vector_allocator_with_stack -> malloc based MVKVector allocator with stack storage
+//
+//////////////////////////////////////////////////////////////////////////////////////////
+template <typename T, int N>
+class mvk_vector_allocator_with_stack

```

Comment:
This appears to be more or less compatible with the `std::allocator` template. In that case, you might be able (for our purposes, anyway) to get away with a templated alias:

```c++
template<typename T>
using MVKVector = std::vector<T, mvk_vector_allocator_with_stack<T, 8>>;
```

You could even add a non-type template parameter for the amount of inline space to reserve:

```c++
template<typename T, size_t N>
using MVKVector = std::vector<T, mvk_vector_allocator_with_stack<T, N>>;
```

Then, for example, the vector of `MTLViewport`s could be made to use a fixed size of 16, so it would never need to reallocate--well, at least as long as the maximum number of viewports remains 16.

Reply:
Did you evaluate the suggestion by @cdavis5e that the allocator might be all you need?